### PR TITLE
fix(server): drive the p2p send and receive halves concurrently

### PR DESCRIPTION
**Pull request by elsiribot**
## Summary

Fixes the p2p deadlock in #9091: two peers that both begin writing a message larger than the transport's flow-control window hang forever, silently. The connected state now drives the send and receive halves as two genuinely concurrent long-lived futures, so a write parked on flow control can no longer stop us from draining the peer. A second commit adds a daily CI job that runs a 19 guardian federation.

## Details

`P2PConnectionStateMachine::transition_connected` multiplexed a single send and a single receive in one `tokio::select!`. Once a peer parked inside `send_message(..).await` — which happens as soon as one message exceeds the window — that task could no longer poll `connection.receive()`. The peer was in the same state for the same reason, so neither drained the other's stream. The connection stayed open, keep-alives kept flowing, `max_idle_timeout` never fired, and nothing was logged on either side: it presents as a federation that quietly stops making progress.

The fix drives a send loop and a receive loop that are never cancelled part-way through a message. That requires the connection to be usable from both halves at once:

- `IP2PConnection::send`/`receive` take `&self` and the trait gains a `Sync` bound.
- `DynP2PConnection` becomes an `Arc`.
- The TLS connection is split into its sink and stream halves behind separate mutexes — one `Framed` cannot serve both directions. Both iroh impls were already `&self`-compatible.

Connection metadata used to be refreshed as a side effect of every message, which long-lived halves no longer provide. Transport changes still arrive on the existing status-update stream; `rtt` gets a 10 s timer. Both refresh **in place** rather than re-entering the connected state, so a send already in flight is never cancelled by a status event.

One variant that looks right and is not: keeping the fused `select!` and draining incoming frames alongside the parked send. That converts a write/write deadlock into a read/read one — each peer blocks in `read_to_end` of the other's message and stops polling its own write. The halves have to be genuinely concurrent, not interleaved within one task.

This is the structural fix, so it also covers the TLS transport, where the same defect bites at ~2.5–3 MB and there is no window knob to turn (`SO_RCVBUF` is capped by `net.core.rmem_max`). Sizing the iroh QUIC windows to `MAX_P2P_MESSAGE_SIZE` is still worth doing on top as defence in depth — it is not in this PR.

## Reviewing

- The `&self` + `Arc` change to `IP2PConnection` is the widest part of the diff and touches the trait every transport implements. Worth checking the TLS split is right: `receive` is still required to be cancel-safe, and `Framed`'s `StreamExt::next` is.
- Refreshing status in place rather than re-entering the connected state is a deliberate behaviour change. It is what keeps a parked send from being cancelled by an unrelated status event, but it means `rtt` is now only as fresh as the 10 s timer instead of per-message. `subscribes_before_snapshot_to_close_status_update_race` was relaxed from `>= 2` subscriptions to `>= 1` for this reason — the connection now holds one subscription for its whole lifetime, which is strictly simpler than re-subscribing per event.
- The daily job runs both p2p stacks sequentially, since `run-with-nix-workspace-ci.sh` takes over the shared target dir. If the runner can afford it, they could go parallel.

## Testing

`receives_while_a_send_is_parked_on_flow_control` is a deterministic regression test built on a fake connection whose `send` only completes once `receive` has been served. It times out against the previous structure and passes against this one — verified both ways.

Also verified end to end against a real devimint federation on the Iroh transport, forcing an all-to-all DKG message of a given size across every peer:

| guardians | payload | before | after |
|---|---|---|---|
| 13 | 1,200,000 B | 13/13 ok | 13/13 ok |
| 13 | 1,300,000 B | **0/13, hung ≥180 s** | 13/13 ok |
| 19 | 1,300,000 B | **0/19, hung ≥170 s** | — |
| 13 | 5,000,000 B | — | 13/13 ok in ~7 s |

The `before` failures reproduce the production symptom exactly: `Still waiting for peer message peer=2`, naming only the lowest missing peer, with zero disconnects and CPU decaying to ~5%.

For context on why this had not been hit at DKG time in practice: unmodified 13/16/19 guardian federations all complete DKG fine (25.7 s / 38.3 s / 61.4 s), with a largest p2p message of 1,750 / 2,134 / 2,518 B. DKG message size is `128·n + 86` bytes — every message is a polynomial commitment of length `threshold`, and mintv2 runs one `run_dkg_g2` per denomination sequentially — so reaching the 1.25 MB window through guardian count alone would take ~9,800 guardians. Payload size is what crosses the threshold, not peer count.

`cargo clippy -p fedimint-server --all-targets` clean, `cargo test -p fedimint-server --lib net::` green (14 tests), `cargo check -p fedimintd -p devimint` clean, `just final-lint` clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01RQkYZiCQyD8QVadFxeLGsN

### CODE REVIEW

**Comment by elsirion** [comment]
@fedimint-bot please review

File: fedimint-server/src/net/p2p.rs [hunk 1]
```diff
@@ -315,60 +322,172 @@ impl<M: Send + 'static> P2PConnectionStateMachine<M> {
...
-                        if self.incoming_sender.try_send(message).is_err() {
-                            debug!(target: LOG_NET_PEER, "Incoming message channel is full");
+                            Some(P2PConnectionSMState::Connected(connection))
                         }
-                    },
-                    Err(e) => return Some(self.disconnect(e)),
-                }
+                    };
+                },
+                e = &mut receive_loop => {
```
Comment (maan2003) [inline fedimint-server/src/net/p2p.rs:378]: why not just put them in real tokio task?

Comment (elsirion) [reply fedimint-server/src/net/p2p.rs:378]: Doesn't seem to provide any benefits over a select

Comment (maan2003) [reply fedimint-server/src/net/p2p.rs:378]: it is harder to deadlock with tasks :shrug:

**Review by maan2003 — APPROVED** [review]
the fix looks correct, this is where agents are much better judging correctness so I trust the bot

**Comment by maan2003** [comment]
@fedimint-bot /review

**Comment by dpc** [comment]
*Posted by Tau*

## Review

Reviewed head `2a21a2cef44f988a9ce9f8b73489ed8053eb58a3`.

**Blocking finding**

- `fedimint-server/src/net/p2p.rs:417-438`: max-age and replacement no longer bound an in-flight send. Once `outgoing_receiver.recv()` returns, `send_loop` waits only on `connection.send(message)`. If the old connection remains transport-alive while the peer stops draining it, that send can stay parked indefinitely; the max-age deadline and a healthy replacement connection are not polled, the state remains connected, and the bounded outgoing queue can start dropping later messages. This removes the recovery path that connection replacement/max-age is meant to provide. Please define the unknown-delivery policy for aborting an in-flight send, then make replacement/max-age able to retire the stuck connection and add regression tests for both cases with a permanently parked send.

The earlier unresolved comment about dropping `receive_loop` during non-cancel-safe `read_to_end()` also still appears valid; I have not duplicated its details here.

I did not run builds or tests as part of this review.

**Review by elsirion — APPROVED** [review]
DPC's comment is more of an edge case and this one has gotten more eyes than an alternative. Happy to follow up should this ever become an actual problem.
